### ISSUE TITLE
[GSoC] Tracking tasks for adding control package

### ISSUE THREAD

**Issue by namannimmo10** [sympy/sympy#19352]
These are the following tasks that need to be completed in order to add a control package to SymPy: 

- [ ]  Adding `StateSpace` class, with basic functionality and unit tests (+ documentation).

- [ ] Adding `TransferFunction` class, with basic functionality and unit tests (+ documentation). 

- [ ] Adding remaining methods using functionality from both classes.

- [ ] Adding `root_locus` plot along with documentation.

- [ ] Adding bode plot along with documentation.

Base PRs:
1. https://github.com/sympy/sympy/pull/18436
2. https://github.com/sympy/sympy/pull/17866

Open PRs:

Merged PRs:

**Comment by namannimmo10** [comment]
ping @moorepants @Sc0rpi0n101 @sylee957 @ishanaj 

Same comment I've had in the PR, if we want to represent the transfer function as a ratio of polynomials, then I think we have to use `Mul`. Since we have decided that `TransferFunction` will subclass from `Basic`, then how can I use `Mul` internally so that: 
```python
tf = TransferFunction((s - 1)*(s - 2), s - 4, s); tf
``` 
gives output: `((s - 2)*(s - 1))/(s - 4)` 
or
``` 
(s - 2)⋅(s - 1)
───────────────
     s - 4 
``` 
with `pprint`. The most basic `__new__` constructor without considering exceptions (like handling non complex symbols, or when denominator is zero) looks like this:
``` python
    def __new__(cls, num, den, var):
        obj = Basic.__new__(cls, num, den, var)
        obj._num = num
        obj._den = den
        obj._var = var
        return obj

    @property
    def num(self):
        return self._num

    @property
    def den(self):
        return self._den

    @property
    def var(self):
        return self._var
```

**Comment by namannimmo10** [comment]
I have added this particular test which uses sympy's own `subs`. And I'm not quite sure whether we should make it work.
```python
tf = TransferFunction(s - 1, s**2 - 2*s + 1, s)
assert tf.subs(s, 2) == TransferFunction(1, 1, s)
```
I think `tf.subs(s, 2)` should raise an Error... cause `TransferFunction(1, 1, 2)`  would be meaningless.

**Comment by moorepants** [comment]
I'm not sure if it is meaningless. I can have meaning in the sense that it is the transfer function evaluated at s=2. We need to support evaluation of the transfer function at specific values of s in some way.

**Comment by moorepants** [comment]
> if we want to represent the transfer function as a ratio of polynomials, then I think we have to use Mul

To answer whether Mul is the right thing to subclass you need to study the behavior of Mul and all of it's methods. If it is clear that you will make use of many of the methods, only overriding some, then Mul may make sense. If not, then you may need to subclass from Mul's parent.

**Comment by oscarbenjamin** [comment]
> I can have meaning in the sense that it is the transfer function evaluated at s=2.

Evaluating the transfer function should give a complex number rather than a transfer function. I wouldn't recommend changing this to return a complex number though as then subs would sometimes return a transfer function and some times an Expr which is inconsistent.

Probably the function argument should be considered a bound symbol as in the case of `Lambda` or `Integral`:
```julia
In [37]: I                                                                                                                                     
Out[37]: 
z     
⌠     
⎮ x dx
⌡     
y     

In [38]: I.subs(x, t)                                                                                                                          
Out[38]: 
z     
⌠     
⎮ x dx
⌡     
y     

In [39]: I.subs(z, t)                                                                                                                          
Out[39]: 
t     
⌠     
⎮ x dx
⌡     
y  

In [40]: I.free_symbols                                                                                                                        
Out[40]: {y, z}
```

**Comment by oscarbenjamin** [comment]
The TransferFunction should not subclass `Mul`. I don't think it should be a subclass of `Expr` (or any `Expr` subclass) either.

**Comment by namannimmo10** [comment]
> The TransferFunction should not subclass `Mul`. I don't think it should be a subclass of `Expr` (or any `Expr` subclass) either.

Any ideas which one should be the superclass?

**Comment by moorepants** [comment]
Start with Basic and see where you get.

**Comment by Sc0rpi0n101** [comment]
Yeah, you should start with subclassing Basic

We can update it later if something comes up.

**Comment by namannimmo10** [comment]
Any ideas which one should be the superclass for both `Series` and `Parallel`?

**Comment by namannimmo10** [comment]
I have a few concerns regarding the implementation:
1. I think we also need a `var` attribute for `Series` and `Parallel` objects. This `var` will be the one that anchors all the transfer functions in a Series/Parallel object. If any transfer function has a different `var`, a `ValueError` is raised.
``` python
>>> tf = TransferFunction(a0*s**3 + a1*s**2 - a2*s, b0*p**4 + b1*p**3 - b2*s*p, s)
>>> tf2 = TransferFunction(a2*p - s, a2*s + p, s)
>>> tf4 = TransferFunction(1, s**2 + 2*zeta*wn*s + wn**2, s)
>>> S1 = Series(tf, tf2, tf4)
>>> S1.var
s
```
2. We have decided that a Series/Parallel object will not support `Expr` in the arguments. We should also not allow any `Function` in the arguments i.e.
```python
>>> inp = Function('X_d')(s)
>>> out = Function('X')(s)
>>> S2 = Series(tf, Parallel(inp, -out), tf2) #: Instead of passing Functions in Parallel object like this.....
```
We can wrap those into `TransferFunction` objects! 
``` python
>>> tf3_ = TransferFunction(inp, 1, s)
>>> tf4_ = TransferFunction(-out, 1, s)
>>> tf5_ = TransferFunction(inp - out, 1, s)
>>> s2 = Series(tf, Parallel(tf3_, tf4_), tf2) #: this works fine...
>>> s3 = Series(tf, Parallel(tf5_, tf4_), tf2) #: this also works fine...
```
Is this fine? Cause everything else will be built on top of it, and I don't want to change the design after implementing it.

**Comment by namannimmo10** [comment]
i see `Function` is an instance of `Expr`, so I think it is fine.

**Comment by ishanaj** [comment]
> We can wrap those into TransferFunction objects!
...
```>>> s2 = Series(tf, Parallel(tf3_, tf4_), tf2)```

Could you also show, what the result of `s2.doit()` would be if we do so? It should conform to the behaviour of functions. (see [comment](https://github.com/sympy/sympy/pull/19390#issuecomment-642758509))
Also, I think wrapping around a function inside a transfer function makes it a composition of functions (function inside a function), which is probably not what you are intending to do.

**Comment by moorepants** [comment]
> I think we also need a var attribute for Series and Parallel objects. 

Yes, in fact Series and Parallel can have all the attributes and methods of TransferFunction and should.

> We can wrap those into TransferFunction objects!

That seems like a good idea. I don't see any issues with that.

**Comment by namannimmo10** [comment]
> Could you also show, what the result of s2.doit() would be if we do so?

``` python
namannimmo@smelly-cat:~/sympy$ ipython --pdb
Python 3.7.3 (default, May 27 2019, 14:03:12) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.15.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from sympy.physics.control.lti import *                                                                                                     

In [2]: from sympy import *; from sympy.abc import *                                                                                                

In [3]: a0, a1, a2, b0, b1, b2 = symbols('a0:3, b0:3')                                                                                              

In [4]: tf = TransferFunction(a0*s**3 + a1*s**2 - a2*s, b0*p**4 + b1*p**3 - b2*s*p, s)                                                              

In [5]: tf2 = TransferFunction(a2*p - s, a2*s + p, s)                                                                                               

In [6]: tf3_ = TransferFunction(Function('X_d')(s), 1, s); tf4_ = TransferFunction(Function('X')(s), 1, s)                                          

In [7]: pprint(Series(tf, Parallel(tf3_, tf4_), tf2).doit())
                           ⎛    3       2       ⎞
(a₂⋅p - s)⋅(X(s) + X_d(s))⋅⎝a₀⋅s  + a₁⋅s  - a₂⋅s⎠
─────────────────────────────────────────────────
                  ⎛    4       3         ⎞       
       (a₂⋅s + p)⋅⎝b₀⋅p  + b₁⋅p  - b₂⋅p⋅s⎠
```

**Comment by moorepants** [comment]
> We can wrap those into TransferFunction objects!

Acutally for this, you often want to separate the numerator and denominator, so you may need to support pure functions of `var` in Parallel and Series.

**Comment by ishanaj** [comment]
I was thinking of `s2.doit()` result to be like: `(X(s) + X_d(s))*TransferFunction(......)`, to make it look simpler.
But I perceive, that's what `Series` and `Parallel` classes are expected to do. (making equations simpler). So, I think that's okay!
You can surely go ahead with it!

**Comment by namannimmo10** [comment]
> Acutally for this, you often want to separate the numerator and denominator, so you may need to support pure functions of `var` in Parallel and Series.

The way I am implementing it, It certainly looks like supporting pure functions of `var` in Parallel and Series will make it unnecessarily complex. The end result would be the same regardless of the way we proceed, so I feel like wrapping those functions into TransferFunction objects is much better.

**Comment by moorepants** [comment]
I don't understand what you mean, so show me the unit tests for the desired behavior and I can comment further.

**Comment by namannimmo10** [comment]
What I am trying to say is that we can support `Expr` arguments in `TransferFunction`, while Series/Parallel objects allow pure functions/ratio of functions wrapped in transfer functions. 
```python
>>> inp1 = Function('X1')(s)
>>> inp2 = Function('X2')(s)
>>> out = Function('X')(s)
>>> tf1 = TransferFunction(out - inp1, out + inp2, s)
>>> tf2 = TransferFunction(a2*p - s, a2*s + p, s)
>>> Series((out - inp1)/(out + inp2), tf2).doit() == Series(tf1, tf2).doit() 
True
>>> Series(tf1, tf2).doit() <--this approach should be followed.
(a₂⋅p - s)⋅(X(s) - X₁(s))
─────────────────────────
(a₂⋅s + p)⋅(X(s) + X₂(s))
```
If we will allow pure functions/ratio of functions in Series/Parallel objects directly, then expressions or numbers could also be included. 
I am using this constructor in Series class.
```python
    def __new__(cls, *args, evaluate=False):
        if not all(isinstance(arg, (TransferFunction, Parallel)) for arg in args):
            raise TypeError

        obj = super(Series, cls).__new__(cls, *args)
        obj._var = None
        for arg in args:
            if obj._var is None:
                obj._var = arg.var
            elif obj._var != arg.var:
                raise ValueError("All transfer functions should be anchored with the same variable.")
        if evaluate:
            return obj.doit()
        return obj
```
 Anyway, I am not quite sure about this, so I will do it either way.

**Comment by namannimmo10** [comment]
I will create a new PR for `TransferFunctionMatrix` class soon.

**Comment by namannimmo10** [comment]
Now, we have two options:
1. We can have a sympy Matrix of transfer functions for MIMO TF:
``` python
>>> tf1 = TransferFunction(1, s**2 + 2*zeta*wn*s + wn**2, s)
>>> tf2 = TransferFunction(k, 1, s)
>>> pprint(Matrix([tf1, tf2]))
⎡         1         ⎤
⎢───────────────────⎥
⎢ 2                2⎥
⎢s  + 2⋅s⋅wn⋅ζ + wn ⎥
⎢                   ⎥
⎢         k         ⎥
⎢         ─         ⎥
⎣         1         ⎦

```
Or, 
2. We can have a new class `TransferFunctionMatrix` for it. This class can also offer some other methods like `is_proper`, `is_strictly_proper`, `is_biproper`, `__neg__`, and `var` so that the user doesn't have to check this manually out of a sympy Matrix.

I think the better option would be to implement a new class for MIMO transfer function.

**Comment by sylee957** [comment]
I also have similar idea to introduce new class for typing issues because the interconnection model also applies to matrix, while matrix operations except for a few may not always be physically realizable for control systems.

But it looks like we have to undergo a hard work to carry over every interconnection models like series, parallel, and feedback again for transfer function matrix.

**Comment by namannimmo10** [comment]
> But it looks like we have to undergo a hard work to carry over every interconnection models like series, parallel, and feedback again for transfer function matrix

Can we make it in-place? A transfer function matrix will only connect with other transfer function matrices (both `series` and `parallel` interconnection), provided they have the same inputs and outputs. 
Although, I am not sure whether MIMO block diagram algebra will work after that.

**Comment by namannimmo10** [comment]
It's already in the `master` branch: 

https://github.com/sympy/sympy/blob/3a812204052176854bdfc1ea9d12097c3f9307ca/sympy/physics/control/lti.py#L915-L925

TIL about CasADi - interesting! I don't know if there's an API to convert sympy matrices to CasADi matrices (only see numpy and scipy in the docs). If you deal with symbolics directly, extracting the expressions and constructing the `casadi` matrix is one option, otherwise sympy matrix -> numpy array -> casadi matrix?

**Comment by billtubbs** [comment]
Indeed it is!  Sorry.  But I don't see it in the docs [here](https://docs.sympy.org/latest/modules/physics/control/lti.html#) yet.

**Comment by billtubbs** [comment]
It looks like I'd have to write a function to parse a sympy expression and construct a new expression using a mapping of sympy symbolic variables to casadi symbolic variables (I don't think that belongs in Sympy but maybe in Casadi).

Any advice on how to do that?  Are there any sympy functions that might be useful to do that?

(I tried this and no surprise that it doesn't work):
```
import casadi as cas

T1_cas = cas.SX.sym('T1')

sys.state_matrix.subs(T1, T1_cas)
```

**Comment by billtubbs** [comment]
Here's a Sympy -> Casadi conversion function if is anyone is interested: 
- https://github.com/casadi/casadi/discussions/3754#discussioncomment-10012864

**Comment by sylee957** [comment]
`sys.state_matrix.subs(T1, T1_cas)` doesn't work because sympy Matrix or sympy expressions are still involved in the middle, and doesn't know how to handle foreign objects in construction.

You already have figured out something to work after the last comment?

You approaches to use lambdify and use mapping is roughly correct.
However, that can be quite indirect approach. If you want to know more about converting sympy objects to other objects, you should first understand the structure of sympy objects
 https://docs.sympy.org/latest/tutorials/intro-tutorial/manipulation.html

**Comment by billtubbs** [comment]
Thanks for the pointers.

I think converting expressions from Sympy to other languages is a whole project in itself.  There are any number of other modelling/numerical optimization packages with their own symbolic languages (CasADi, Acados, Neuromancer, Open Modelica, Dymola, GEKKO, Pyomo, MATLAB, ... just the ones I know).  Correct me if I'm wrong but I assume Sympy is far more sophisticated than any of these because it is a dedicated, general purpose symbolic analysis library.  The fact that you are adding control analysis tools reinforces that.

While it would be a big project, I think it would be a great if there were more symbolic language conversion tools so that symbolic manipulation and analysis in Python could be centralized in one place (Sympy) and then the results ported into whatever end-use tool the user wants to use. (As opposed to every other library extending it's symbolic language capabilities).

Has any work been done on this?  (In this project or elsewhere)

**Comment by billtubbs** [comment]
Sorry, I know this is a bit off-topic for this discussion.  Is there a better place where we could compile a status list of symbolic language converters (i.e. sympy_to_XYZ and XYZ_to_sympy)?

**Comment by sylee957** [comment]
> Sorry, I know this is a bit off-topic for this discussion. Is there a better place where we could compile a status list of symbolic language converters (i.e. sympy_to_XYZ and XYZ_to_sympy)?

We can add some compatibility lists in the documentation to get overview, that can be helpful, however, I still worry about if such things can get outdated without maintainace.
So if you search some keywords in the code or documentation and don't find anything there, it's likely that they don't exist in sympy core yet. We appreciate to answer such questions because for now and you don't need to worry about coming with questions that can be duplicate. In fact, `CasADi` is quite new to hear.

However, among your list (CasADi, Acados, Neuromancer, Open Modelica, Dymola, GEKKO, Pyomo, MATLAB), 
I think that only MATLAB is supported a bit indirectly with Octave (see https://docs.sympy.org/latest/modules/printing.html#module-sympy.printing.octave)
However, sympy is only compatible with MATLAB for general mathematical expressions. Matlab is quite big and in terms of individual toolboxes (symbolic toolbox, or control toolbox) there can be something missing or there can be something difficult to implement due to the advanced features in Matlab.
